dal/dal.go

```diff
@@ -116,22 +120,101 @@ func (d *Dal) KeyExists(key string) (bool, bool, error) {
 	return true, resp.Node.Dir, nil
 }
 
-// An unwieldy wrapper for setting a new key
-func (d *Dal) Set(key, value string, dir bool, ttl int, prevExist string) error {
-	existState := client.PrevExistType(prevExist)
+type SetOptions struct {
+	// If SetOptions.Dir=true then value is ignored.
+	Dir       bool
+	TTLSec    int
+	PrevExist string
+
+	// Create parents will recursively create parent directories as needed.
+	CreateParents bool
+
+	// To be used with create parents.
+	// If depth > 0 it will only create depth number of parents.
+	// If depth < 0 it will try to create as many parents as necessary.
+	// If depth == 0 it will not create any parents. This behaves the same as Set().
+	Depth int
+}
+
+// A wrapper for setting a new key
+func (d *Dal) Set(key, value string, opt *SetOptions) error {
+	if key[0] != '/' {
+		key = "/" + key
+	}
+
+	// enforce this just in case
+	if !opt.CreateParents {
+		opt.Depth = 0
+	}
+
+	//TODO handle the case that depth is lager than the num of path elements
+
+	err := d.setAndCreateParents(
+		d.Prefix+key,
+		value,
+		opt.Dir,
+		opt.Depth,
+		time.Duration(opt.TTLSec)*time.Second,
+		client.PrevExistType(opt.PrevExist),
+	)
+
+	return err
+}
+
+// Helper method for set to recursively create its parents if they do not exist.
+// It will create up to `depth` number of parents.
+func (d *Dal) setAndCreateParents(
+	key, value string, dir bool, depth int, ttl time.Duration, pExist client.PrevExistType) error {
 
 	_, err := d.KeysAPI.Set(
 		context.Background(),
-		d.Prefix+"/"+key,
+		key,
 		value,
 		&client.SetOptions{
 			Dir:       dir,
-			TTL:       time.Duration(ttl) * time.Second,
-			PrevExist: existState,
+			TTL:       ttl,
+			PrevExist: pExist,
 		},
 	)
 
-	return err
+	if err != nil {
+		// is an etcd error
+		etcdErr, ok := err.(client.Error)
+		if !ok {
+			return err
+		}
+
+		// parent creation is enabled, and error is a key not found
+		if depth != 0 && etcdErr.Code == client.ErrorCodeKeyNotFound {
+			// recursively create its parent first
+			parent := key[:strings.LastIndex(key, "/")] //TODO should use path lib to manipulate this
+			if err := d.setAndCreateParents(parent, "", true, depth-1, ttl, client.PrevIgnore); err != nil {
+				return err
+			}
+
+			// now try to set it again
+			if _, err := d.KeysAPI.Set(
+				context.Background(),
+				key,
+				value,
+				&client.SetOptions{
+					Dir:       dir,
+					TTL:       ttl,
+					PrevExist: pExist,
+				},
+			); err != nil {
+				// give up if that didnt work
+				return err
```

Comment:
Let's make this more explicit - ie. `fmt.Errorf("Failed to complete parent dir creation: %v", err)` -- that way it's a bit more obvious as to what happened.